rust-mode.el

```diff
@@ -267,18 +268,40 @@ to the function arguments.  When nil, `->' will be indented one level."
         ;; Rewind until the point no longer moves
         (setq continue (/= starting (point)))))))
 
-(defun rust-in-macro ()
+(defvar-local rust-macro-scopes nil)
+
+(defun rust-macro-scope (&optional start end)
   (save-excursion
-    (when (> (rust-paren-level) 0)
-      (backward-up-list)
-      (rust-rewind-irrelevant)
-      (or (rust-looking-back-macro)
-          (and (rust-looking-back-ident)
-               (save-excursion
-                 (backward-sexp)
-                 (rust-rewind-irrelevant)
-                 (rust-looking-back-str "macro_rules!")))
-          (rust-in-macro)))))
+    ;; need to special case macro_rules which has unique syntax
+    (let ((scope nil))
+      (goto-char
+       (or start (point-min)))
+      ;; if there is a start move back to the first top level
+      (when start
+        (let ((top (syntax-ppss-toplevel-pos (syntax-ppss))))
+          (when top
+            (goto-char top))))
+      (while (search-forward "!" end t)
+        (when (and (not (rust-in-str-or-cmnt))
+                   (skip-chars-forward " \t\n\r")
+                   (or
+                    (eq ?\[ (char-after))
+                    (eq ?\( (char-after))
+                    (eq ?\{ (char-after))))
+          (let ((start (point)))
+            (forward-list)
+            (setq scope (cons (list start (point)) scope)))))
+      scope)))
```

Comment:
Here if we are in a region without macros, scope is nil, which means that rust-in-macro will call rust-macro-scopes repeatedly. Better return `(cons scope nil)` and then in rust-in-macro use `(car scopes)` after the check for nil.

Reply:
I don't understand. `rust-in-macro` should always call `rust-macro-scope`, unless `rust-macro-scopes` is `let`-bound which it is in `rust-syntax-propertize`. That's the only place `rust-in-macro` is called a repeatedly within a single turn of the command-loop.

Reply:
Let's say I have a large file with not a single exclamation mark (or the region between start and end, doesn't have to be the whole file). Then rust-macro-scope sets rust-macro-scopes to nil as there are no macro scopes. Then rust-in-macro sees that rust-macro-scopes is nil so it calls rust-macro-scope every time.

Reply:
Ah, yes, good point. Fixed this now, but in a slightly different way.

Reply:
It's still broken. The `defvar-local` should still be to nil as no pass has been done yet. And in rust-in-macro, scopes should still be bound to `(or rust-macro-scopes (rust-macro-scope ...))` to avoid a repeated search on `'empty`.

Reply:
Oh, dear. I am missing my own "assess" library which makes testing this sort of thing easier. I believe I have fixed this properly now.